RFC-0025-improving-incremental-builds.md

```diff
@@ -0,0 +1,217 @@
+#include <ATen/ops/add_native.h> // Split from NativeFunctions.h
+#include <ATen/ops/add_ops.h>    // Split from Operators.h
+
+// Plus static dispatch functions split from {DispatchKey}Functions.h
+#include <ATen/ops/add_compositeexplicitautograd_dispatch.h>
+#include <ATen/ops/add_cpu_dispatch.h>
+#include <ATen/ops/add_cuda_dispatch.h>
+#include <ATen/ops/add_meta_dispatch.h>
+```
+
+Note that `add` has 4 overloads which are included in the same header
+files. This way the caller doesn't need to know which overload is
+relevant to their use case, only the basic operator name.
+
+Unfortunately, per-operator headers aren't compatible with Meta's
+internal `buck` builds because the codegen output list depends
+dynamically on the contents of an input file (namely
+`native_functions.yaml`). So these headers can only be included when
+`AT_PER_OPERATOR_HEADERS` is defined which is the default for CMake
+and bazel builds. So, they must be included behind an `#ifdef` guard,
+e.g.
+
+```cpp
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#else
+#include <ATen/ops/add.h>
+#include <ATen/ops/sum.h>
+#endif
+```
+
+### Enforcing adoption
+These tools can only improve incremental builds if they are used
+widely throughout the codebase and continue to be enforced as the code
+develops. So, adoption will be enforced through the
+`TORCH_ASSERT_NO_OPERATORS` and `TORCH_ASSERT_ONLY_METHOD_OPERATORS`
+macros.
+
+First, `ATen` and `torch` will be migrated to using the per-operator
+headers, defining `TORCH_ASSERT_ONLY_METHOD_OPERATORS` in each file as
+they are updated. Then, once most of the codebase is compliant, the
+build system will be changed to define
+`TORCH_ASSERT_ONLY_METHOD_OPERATORS` by default in all new files. In
+the rare occasion where umbrella headers are strictly necessary, these
+files can opt-out by explicitly `#undef`ing the macro making it
+`grep`-able and clearly visible in code-review. One such example is
+`autograd/generated/Functions.cpp` which includes verbatim code from
+`derivatives.yaml` so doesn't know which operators are being called.
+This should be extremely rare though.
+
+The initial migration can be assisted by
+[`include-what-you-use`](https://include-what-you-use.org/) which is a
+clang-based tool to find the necessary headers based on what symbols
+are actually used in a file. For most cpp files, it is as simple as
+copying the `<ATen/ops/*>` headers from the tool output.
+
+`TORCH_ASSERT_NO_OPERATORS` cannot be enforced as strictly because
+compliance often requires significantly restructuring the code.
+However, where the cost-benefit is favorable, this macro should be
+defined manually in the source file. This will include files that
+already use `TensorIterator` exclusively and just need their includes
+pruned; and more costly to compile files where taking time to
+restructure them is worthwhile.
+
+## **Metrics**
+
+Some key metrics that should be improved by these changes:
```

Comment:
Do we have some measure of these metrics. Or a snapshot of todays values so that we can measure future improvements.

Reply:
I've added a table with some measurements. The "after" measurements are incomplete of course because I'm not finished yet and much of ATen and torch still include `ATen.h`.

Reply:
I think it's good to keep updating that table so that we have a place with the historical data and improvements as we move forward.